src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +314,515 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
+	 *       only bother non-colocated subquery joins, which only happens when
+	 *       there are subqueries.
+	 */
+	if (SubqueryEntryList(subquery) == NIL && SublinkList(subquery) == NIL)
+	{
+		return false;
+	}
+
+	/* direct joins with local tables are not supported by any of Citus planners */
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		return false;
+	}
+
+	/*
+	 * Finally, check whether this subquery contains distribution key equality or not.
+	 */
+	if (!SubqueryContainsDistributionKeyEquality(subquery,
+												 context->plannerRestrictionContext))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * RecursivelyPlanNonColocatedSubqueries gets a query which includes one or more
+ * other subqueries that are not joined on their distribution keys. The function
+ * tries to recursively plan some of the subqueries to make the input query
+ * executable by Citus.
+ *
+ * The function picks an anchor subquery and iterates on the remaining subqueries.
+ * Whenever it finds a non colocated subquery with the anchor subquery, the function
+ * decides to recursively plan the non colocated subquery.
+ *
+ * The function first handles subqueries in FROM clause (i.e., jointree->fromlist) and then
+ * subqueries in WHERE clause (i.e., jointree->quals).
+ *
+ * The function does not treat outer joins seperately. Thus, we might end up with
+ * a query where the function decides to recursively plan an outer side of an outer
+ * join (i.e., LEFT side of LEFT JOIN). For simplicity, we chose to do so and handle
+ * outer joins with a seperate pass on the join tree.
+ */
+static void
+RecursivelyPlanNonColocatedSubqueries(Query *subquery, RecursivePlanningContext *context)
+{
+	NonColocatedSubqueryContext nonColocatedPlanningContext;
+
+	RangeTblEntry *anchorRangeTblEntry = NULL;
+
+	FromExpr *joinTree = subquery->jointree;
+	PlannerRestrictionContext *restrictionContext = NULL;
+
+	/* we couldn't pick an anchor subquery, so we cannot continue */
+	anchorRangeTblEntry = AnchorRte(subquery);
+	if (anchorRangeTblEntry == NULL)
+	{
+		return;
+	}
+
+	/* create the context for the non colocated subquery planning */
+	restrictionContext = context->plannerRestrictionContext;
+	nonColocatedPlanningContext =
+		CreateNonColocatedSubqueryJoinContext(subquery, anchorRangeTblEntry,
+											  restrictionContext);
+
+	/* handle from clause subqueries first */
+	RecursivelyPlanNonColocatedJoinWalker((Node *) joinTree, &nonColocatedPlanningContext,
+										  context);
+
+	/* handle subqueries in WHERE clause */
+	RecursivelyPlanNonColocatedSubqueriesInWhere(subquery, &nonColocatedPlanningContext,
+												 context);
+}
+
+
+/*
+ * CreateNonColocatedSubqueryJoinContext is a helper function that simply calculates
+ * a NonColocatedSubqueryContext with the given input and returns it.
+ */
+static NonColocatedSubqueryContext
+CreateNonColocatedSubqueryJoinContext(Query *subquery, RangeTblEntry *anchorRangeTblEntry,
```

Comment:
This file is getting  big. Would it be possible to move some of these functions into a different file that we can eventually expand with more advanced logic? e.g. `noncolocated_join_planning.c`

Reply:
Hmm, that's a fair point. 

The main problem I see is how do we pass `RecursivePlanningContext ` struct to the lower level functions? I think we don't want to make it public as well. And, without making that public, the benefit of moving into a separate file seems low since half of the functions use that.

What if we create a new header file (e.g., 'recursive_planning_typedefs.h'), which only includes the `RecursivePlanningContext` struct, and we comment that it should only be included by recursive planning source files?